File: Raven.Client.Lightweight/Indexes/IndexDefinitionBuilder.cs
//-----------------------------------------------------------------------
// <copyright file="IndexDefinition.cs" company="Hibernating Rhinos LTD">
//     Copyright (c) Hibernating Rhinos LTD. All rights reserved.
// </copyright>
//-----------------------------------------------------------------------
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq.Expressions;
using Raven.Abstractions.Extensions;
using Raven.Abstractions.Indexing;
using Raven.Client.Document;

namespace Raven.Client.Indexes
{
	/// <summary>
	/// This class provides a way to define a strongly typed index on the client.
	/// </summary>
	public class IndexDefinitionBuilder<TDocument, TReduceResult>
	{
		/// <summary>
		/// Gets or sets the map function
		/// </summary>
		/// <value>The map.</value>
		public Expression<Func<IEnumerable<TDocument>, IEnumerable>> Map { get; set; }
		/// <summary>
		/// Gets or sets the reduce function
		/// </summary>
		/// <value>The reduce.</value>
		public Expression<Func<IEnumerable<TReduceResult>, IEnumerable>> Reduce { get; set; }

		/// <summary>
		/// Gets or sets the reduce function
		/// </summary>
		/// <value>The reduce.</value>
		[Obsolete("Use Result Transformers instead.")]
		public Expression<Func<IClientSideDatabase, IEnumerable<TReduceResult>, IEnumerable>> TransformResults { get; set; }

		/// <summary>
		/// Gets or sets the stores options
		/// </summary>
		/// <value>The stores.</value>
		public IDictionary<Expression<Func<TReduceResult, object>>, FieldStorage> Stores { get; set; }

		/// <summary>
		/// Gets or sets the stores options
		/// </summary>
		/// <value>The stores.</value>
		public IDictionary<string, FieldStorage> StoresStrings { get; set; }

		/// <summary>
		/// Gets or sets the indexing options
		/// </summary>
		/// <value>The indexes.</value>
		public IDictionary<Expression<Func<TReduceResult, object>>, FieldIndexing> Indexes { get; set; }

		/// <summary>
		/// Gets or sets the indexing options
		/// </summary>
		/// <value>The indexes.</value>
		public IDictionary<string, FieldIndexing> IndexesStrings { get; set; }

		/// <summary>
		/// Gets or sets the sort options.
		/// </summary>
		/// <value>The sort options.</value>
		public IDictionary<Expression<Func<TReduceResult, object>>, SortOptions> SortOptions { get; set; }

		/// <summary>
        /// Gets or sets the sort options.
        /// </summary>
        /// <value>The sort options.</value>
        public Dictionary<string, SortOptions> SortOptionsStrings { get; set; }

		/// <summary>
		/// Get os set the analyzers
		/// </summary>
		public IDictionary<Expression<Func<TReduceResult, object>>, string> Analyzers { get; set; }

		/// <summary>
		/// Get os set the analyzers
		/// </summary>
		public IDictionary<string, string> AnalyzersStrings { get; set; }

		/// <summary>
		/// Gets or sets the suggestion options.
		/// </summary>
		/// <value>The suggestion options.</value>
		public IDictionary<Expression<Func<TReduceResult, object>>, SuggestionOptions> Suggestions { get; set; }

		/// <summary>
		/// Gets or sets the term vector options
		/// </summary>
		/// <value>The term vectors.</value>
		public IDictionary<Expression<Func<TReduceResult, object>>, FieldTermVector> TermVectors { get; set; }

		/// <summary>
		/// Gets or sets the term vector options
		/// </summary>
		/// <value>The term vectors.</value>
		public IDictionary<string, FieldTermVector> TermVectorsStrings { get; set; }

		/// <summary>
		/// Gets or sets the spatial options
		/// </summary>
		/// <value>The spatial options.</value>
		public IDictionary<Expression<Func<TReduceResult, object>>, SpatialOptions> SpatialIndexes { get; set; }

		/// <summary>
		/// Gets or sets the spatial options
		/// </summary>
		/// <value>The spatial options.</value>
		public IDictionary<string, SpatialOptions> SpatialIndexesStrings { get; set; }

		/// <summary>
		/// Max number of allowed indexing outputs per one source document
		/// </summary>
		public int? MaxIndexOutputsPerDocument { get; set; }

		/// <summary>
		/// Initializes a new instance of the <see cref="IndexDefinitionBuilder{TDocument,TReduceResult}"/> class.
		/// </summary>
		public IndexDefinitionBuilder()
		{
			Stores = new Dictionary<Expression<Func<TReduceResult, object>>, FieldStorage>();
			StoresStrings = new Dictionary<string, FieldStorage>();
			Indexes = new Dictionary<Expression<Func<TReduceResult, object>>, FieldIndexing>();
			IndexesStrings = new Dictionary<string, FieldIndexing>();
			SortOptions = new Dictionary<Expression<Func<TReduceResult, object>>, SortOptions>();
            SortOptionsStrings = new Dictionary<string, SortOptions>();
			Suggestions = new Dictionary<Expression<Func<TReduceResult, object>>, SuggestionOptions>();
			Analyzers = new Dictionary<Expression<Func<TReduceResult, object>>, string>();
			AnalyzersStrings = new Dictionary<string, string>();
			TermVectors = new Dictionary<Expression<Func<TReduceResult, object>>, FieldTermVector>();
			TermVectorsStrings = new Dictionary<string, FieldTermVector>();
			SpatialIndexes = new Dictionary<Expression<Func<TReduceResult, object>>, SpatialOptions>();
			SpatialIndexesStrings = new Dictionary<string, SpatialOptions>();
		}

		/// <summary>
		/// Toes the index definition.
		/// </summary>
		public IndexDefinition ToIndexDefinition(DocumentConvention convention, bool validateMap = true)
		{
			if (Map == null && validateMap)
				throw new InvalidOperationException(
					string.Format("Map is required to generate an index, you cannot create an index without a valid Map property (in index {0}).", GetType().Name));

			if (Reduce != null)
				IndexDefinitionHelper.ValidateReduce(Reduce);

			string querySource = (typeof(TDocument) == typeof(object) || ContainsWhereEntityIs()) ? "docs" : "docs." + convention.GetTypeTagName(typeof(TDocument));
			var indexDefinition = new IndexDefinition
			{
				Reduce = IndexDefinitionHelper.PruneToFailureLinqQueryAsStringToWorkableCode<TDocument, TReduceResult>(Reduce, convention, "results", translateIdentityProperty: false),
#pragma warning disable 612,618
				TransformResults = IndexDefinitionHelper.PruneToFailureLinqQueryAsStringToWorkableCode<TDocument, TReduceResult>(TransformResults, convention, "results", translateIdentityProperty: Reduce == null),
#pragma warning restore 612,618
				Indexes = ConvertToStringDictionary(Indexes),
				Stores = ConvertToStringDictionary(Stores),
				SortOptions = ConvertToStringDictionary(SortOptions),
				Analyzers = ConvertToStringDictionary(Analyzers),
				Suggestions = ConvertToStringDictionary(Suggestions),
				TermVectors =  ConvertToStringDictionary(TermVectors),
				SpatialIndexes = ConvertToStringDictionary(SpatialIndexes),
				MaxIndexOutputsPerDocument = MaxIndexOutputsPerDocument
			};

			foreach (var indexesString in IndexesStrings)
			{
				if (indexDefinition.Indexes.ContainsKey(indexesString.Key))
					throw new InvalidOperationException("There is a duplicate key in indexes: " + indexesString.Key);
				indexDefinition.Indexes.Add(indexesString);
			}

			foreach (var storeString in StoresStrings)
			{
				if (indexDefinition.Stores.ContainsKey(storeString.Key))
					throw new InvalidOperationException("There is a duplicate key in stores: " + storeString.Key);
				indexDefinition.Stores.Add(storeString);
			}

			foreach (var analyzerString in AnalyzersStrings)
			{
				if (indexDefinition.Analyzers.ContainsKey(analyzerString.Key))
					throw new InvalidOperationException("There is a duplicate key in analyzers: " + analyzerString.Key);
				indexDefinition.Analyzers.Add(analyzerString);
			}

			foreach (var termVectorString in TermVectorsStrings)
			{
				if (indexDefinition.TermVectors.ContainsKey(termVectorString.Key))
					throw new InvalidOperationException("There is a duplicate key in term vectors: " + termVectorString.Key);
				indexDefinition.TermVectors.Add(termVectorString);
			}

			foreach (var spatialString in SpatialIndexesStrings)
			{
				if (indexDefinition.SpatialIndexes.ContainsKey(spatialString.Key))
					throw new InvalidOperationException("There is a duplicate key in spatial indexes: " + spatialString.Key);
				indexDefinition.SpatialIndexes.Add(spatialString);
			}

            foreach (var sortOption in SortOptionsStrings)
            {
                if (indexDefinition.SortOptions.ContainsKey(sortOption.Key))
                    throw new InvalidOperationException("There is a duplicate key in sort options: " + sortOption.Key);
                indexDefinition.SortOptions.Add(sortOption);
            }

			if (Map != null)
				indexDefinition.Map = IndexDefinitionHelper.PruneToFailureLinqQueryAsStringToWorkableCode<TDocument, TReduceResult>(Map, convention,
																													 querySource,
																													 translateIdentityProperty
																														: true);

			return indexDefinition;
		}

		private bool ContainsWhereEntityIs()
		{
		    if (Map == null) return false;
			var whereEntityIsVisitor = new WhereEntityIsVisitor();
			whereEntityIsVisitor.Visit(Map.Body);
			return whereEntityIsVisitor.HasWhereEntityIs;
		}

		private class WhereEntityIsVisitor : ExpressionVisitor
		{
			public bool HasWhereEntityIs { get; set; }

			protected override Expression VisitMethodCall(MethodCallExpression node)
			{
				if (node.Method.Name == "WhereEntityIs")
					HasWhereEntityIs = true;
				return base.VisitMethodCall(node);
			}
		}

		private static IDictionary<string, TValue> ConvertToStringDictionary<TValue>(IEnumerable<KeyValuePair<Expression<Func<TReduceResult, object>>, TValue>> input)
		{
			var result = new Dictionary<string, TValue>();
			foreach (var value in input)
			{
				var propertyPath = value.Key.ToPropertyPath('_');
				result[propertyPath] = value.Value;
			}
			return result;
		}
	}

	/// <summary>
	/// This class provides a way to define a strongly typed index on the client.
	/// </summary>
	public class IndexDefinitionBuilder<TDocument> : IndexDefinitionBuilder<TDocument, TDocument> { }
}

File: Raven.Client.Java/src/main/java/net/ravendb/client/indexes/IndexDefinitionBuilder.java
package net.ravendb.client.indexes;

import java.util.HashMap;
import java.util.Map;
import java.util.Map.Entry;

import net.ravendb.abstractions.extensions.ExpressionExtensions;
import net.ravendb.abstractions.indexing.FieldIndexing;
import net.ravendb.abstractions.indexing.FieldStorage;
import net.ravendb.abstractions.indexing.FieldTermVector;
import net.ravendb.abstractions.indexing.IndexDefinition;
import net.ravendb.abstractions.indexing.SortOptions;
import net.ravendb.abstractions.indexing.SpatialOptions;
import net.ravendb.abstractions.indexing.SuggestionOptions;
import net.ravendb.client.document.DocumentConvention;


import com.mysema.query.types.Path;



public class IndexDefinitionBuilder {

  private String map;
  private String reduce;
  private String transformResults;
  private Map<Path<?>, FieldStorage> stores;
  private Map<String, FieldStorage> storesStrings;

  private Long maxIndexOutputsPerDocument;
  private boolean disableInMemoryIndexing;

  private Map<Path<?>, FieldIndexing> indexes;
  private Map<String, FieldIndexing> indexesStrings;

  private Map<Path<?>, SortOptions> sortOptions;
  private Map<String, SortOptions> sortOptionsStrings;
  private Map<Path<?>, String> analyzers;
  private Map<String, String> analyzersStrings;
  private Map<Path<?>, SuggestionOptions> suggestions;
  private Map<Path<?>, FieldTermVector> termVectors;
  private Map<String, FieldTermVector> termVectorsStrings;
  private Map<Path<?>, SpatialOptions> spatialIndexes;
  private Map<String, SpatialOptions> spatialIndexesStrings;


  public IndexDefinitionBuilder() {
    this.stores = new HashMap<>();
    this.storesStrings = new HashMap<>();
    this.indexes = new HashMap<>();
    this.indexesStrings = new HashMap<>();
    this.sortOptions = new HashMap<>();
    this.sortOptionsStrings = new HashMap<>();
    this.suggestions = new HashMap<>();
    this.analyzers = new HashMap<>();
    this.analyzersStrings = new HashMap<>();
    this.termVectors = new HashMap<>();
    this.termVectorsStrings = new HashMap<>();
    this.spatialIndexes = new HashMap<>();
    this.spatialIndexesStrings = new HashMap<>();
  }

  private <S> Map<String, S> convertToStringDictionary(Map<Path< ? >, S> input) {
    Map<String, S> result = new HashMap<>();
    for (Entry<Path<?>, S> value: input.entrySet()) {
      String propertyPath = ExpressionExtensions.toPropertyPath(value.getKey(), '_');
      result.put(propertyPath, value.getValue());
    }
    return result;
  }
  public Map<Path< ? >, String> getAnalyzers() {
    return analyzers;
  }
  public Map<String, String> getAnalyzersStrings() {
    return analyzersStrings;
  }
  public Map<Path< ? >, FieldIndexing> getIndexes() {
    return indexes;
  }
  public Map<String, FieldIndexing> getIndexesStrings() {
    return indexesStrings;
  }
  public String getMap() {
    return map;
  }
  public String getReduce() {
    return reduce;
  }

  public boolean isDisableInMemoryIndexing() {
    return disableInMemoryIndexing;
  }


  public void setDisableInMemoryIndexing(boolean disableInMemoryIndexing) {
    this.disableInMemoryIndexing = disableInMemoryIndexing;
  }

  public Map<Path< ? >, SortOptions> getSortOptions() {
    return sortOptions;
  }
  public Map<Path< ? >, SpatialOptions> getSpatialIndexes() {
    return spatialIndexes;
  }
  public Map<String, SpatialOptions> getSpatialIndexesStrings() {
    return spatialIndexesStrings;
  }
  public Map<Path< ? >, FieldStorage> getStores() {
    return stores;
  }
  public Map<String, FieldStorage> getStoresStrings() {
    return storesStrings;
  }
  public Map<Path< ? >, SuggestionOptions> getSuggestions() {
    return suggestions;
  }
  public Map<Path< ? >, FieldTermVector> getTermVectors() {
    return termVectors;
  }
  public Map<String, FieldTermVector> getTermVectorsStrings() {
    return termVectorsStrings;
  }
  public Map<String, SortOptions> getSortOptionsStrings() {
    return sortOptionsStrings;
  }
  public void setSortOptionsStrings(Map<String, SortOptions> sortOptionsStrings) {
    this.sortOptionsStrings = sortOptionsStrings;
  }
  public Long getMaxIndexOutputsPerDocument() {
    return maxIndexOutputsPerDocument;
  }
  public void setMaxIndexOutputsPerDocument(Long maxIndexOutputsPerDocument) {
    this.maxIndexOutputsPerDocument = maxIndexOutputsPerDocument;
  }

  public String getTransformResults() {
    return transformResults;
  }

  public void setAnalyzers(Map<Path< ? >, String> analyzers) {
    this.analyzers = analyzers;
  }
  public void setAnalyzersStrings(Map<String, String> analyzersStrings) {
    this.analyzersStrings = analyzersStrings;
  }
  public void setIndexes(Map<Path< ? >, FieldIndexing> indexes) {
    this.indexes = indexes;
  }
  public void setIndexesStrings(Map<String, FieldIndexing> indexesStrings) {
    this.indexesStrings = indexesStrings;
  }
  public void setMap(String map) {
    this.map = map;
  }
  public void setReduce(String reduce) {
    this.reduce = reduce;
  }
  public void setSortOptions(Map<Path< ? >, SortOptions> sortOptions) {
    this.sortOptions = sortOptions;
  }
  public void setSpatialIndexes(Map<Path< ? >, SpatialOptions> spatialIndexes) {
    this.spatialIndexes = spatialIndexes;
  }
  public void setSpatialIndexesStrings(Map<String, SpatialOptions> spatialIndexesStrings) {
    this.spatialIndexesStrings = spatialIndexesStrings;
  }
  public void setStores(Map<Path< ? >, FieldStorage> stores) {
    this.stores = stores;
  }
  public void setStoresStrings(Map<String, FieldStorage> storesStrings) {
    this.storesStrings = storesStrings;
  }
  public void setSuggestions(Map<Path< ? >, SuggestionOptions> suggestions) {
    this.suggestions = suggestions;
  }

  public void setTermVectors(Map<Path< ? >, FieldTermVector> termVectors) {
    this.termVectors = termVectors;
  }

  public void setTermVectorsStrings(Map<String, FieldTermVector> termVectorsStrings) {
    this.termVectorsStrings = termVectorsStrings;
  }

  public void setTransformResults(String transformResults) {
    this.transformResults = transformResults;
  }

  public IndexDefinition toIndexDefinition(DocumentConvention convention) {
    return toIndexDefinition(convention, true);
  }

  @SuppressWarnings("deprecation")
  public IndexDefinition toIndexDefinition(DocumentConvention convention, boolean validateMap) {
    if (map == null && validateMap)
      throw new IllegalStateException(
        String.format("Map is required to generate an index, you cannot create an index without a valid Map property (in index %s).", getClass().getSimpleName()));

    IndexDefinition indexDefinition = new IndexDefinition();
    if (reduce != null) {
      indexDefinition.setReduce(reduce);
    }
    if (transformResults != null) {
      indexDefinition.setTransformResults(transformResults);
    }
    indexDefinition.setIndexes(convertToStringDictionary(indexes));
    indexDefinition.setStores(convertToStringDictionary(stores));
    indexDefinition.setSortOptions(convertToStringDictionary(sortOptions));
    indexDefinition.setAnalyzers(convertToStringDictionary(analyzers));
    indexDefinition.setSuggestions(convertToStringDictionary(suggestions));
    indexDefinition.setTermVectors(convertToStringDictionary(termVectors));
    indexDefinition.setSpatialIndexes(convertToStringDictionary(spatialIndexes));

    for (Map.Entry<String, FieldIndexing> indexesString: indexesStrings.entrySet()) {
      if (indexDefinition.getIndexes().containsKey(indexesString.getKey())) {
        throw new IllegalArgumentException("There is a duplicate key in indexes: " + indexesString.getKey());
      }
      indexDefinition.getIndexes().put(indexesString.getKey(), indexesString.getValue());
    }

    for (Map.Entry<String, FieldStorage> storeString: storesStrings.entrySet()) {
      if (indexDefinition.getStores().containsKey(storeString.getKey()))
        throw new IllegalArgumentException("There is a duplicate key in stores: " + storeString.getKey());
      indexDefinition.getStores().put(storeString.getKey(), storeString.getValue());
    }

    for (Map.Entry<String, String> analyzerString: analyzersStrings.entrySet()) {
      if (indexDefinition.getAnalyzers().containsKey(analyzerString.getKey()))
        throw new IllegalArgumentException("There is a duplicate key in analyzers: " + analyzerString.getKey());
      indexDefinition.getAnalyzers().put(analyzerString.getKey(), analyzerString.getValue());
    }

    for (Map.Entry<String, FieldTermVector> termVectorString: termVectorsStrings.entrySet()) {
      if (indexDefinition.getTermVectors().containsKey(termVectorString.getKey()))
        throw new IllegalArgumentException("There is a duplicate key in term vectors: " + termVectorString.getKey());
      indexDefinition.getTermVectors().put(termVectorString.getKey(), termVectorString.getValue());
    }

    for (Map.Entry<String, SpatialOptions> spatialString: spatialIndexesStrings.entrySet()) {
      if (indexDefinition.getSpatialIndexes().containsKey(spatialString.getKey()))
        throw new IllegalArgumentException("There is a duplicate key in spatial indexes: " + spatialString.getKey());
      indexDefinition.getSpatialIndexes().put(spatialString.getKey(), spatialString.getValue());
    }

    for (Map.Entry<String, SortOptions> sortOptions: sortOptionsStrings.entrySet()) {
      if (indexDefinition.getSortOptions().containsKey(sortOptions.getKey()))
        throw new IllegalArgumentException("There is a duplicate key in sort options: " + sortOptions.getKey());
      indexDefinition.getSortOptions().put(sortOptions.getKey(), sortOptions.getValue());
    }

    indexDefinition.setDisableInMemoryIndexing(disableInMemoryIndexing);
    indexDefinition.setMaxIndexOutputsPerDocument(maxIndexOutputsPerDocument);

    if (map != null) {
      indexDefinition.setMap(map);
    }

    return indexDefinition;
  }



}
